src/implementations/title-escrow/helpers.ts

```diff
@@ -0,0 +1,71 @@
+import { TitleEscrowFactory, TradeTrustErc721Factory } from "@govtechsg/token-registry";
+import { Wallet } from "ethers";
+import signale from "signale";
+import { ConnectedSigner } from "../utils/wallet";
+
+interface ConnectToTitleEscrowArgs {
+  tokenId: string;
+  address: string;
+  wallet: Wallet | ConnectedSigner;
+}
+
+type TitleEscrowInstanceType = ReturnType<typeof TitleEscrowFactory.connect>;
+
+export const connectToTitleEscrow = async ({
+  tokenId,
+  address,
+  wallet,
+}: ConnectToTitleEscrowArgs): Promise<TitleEscrowInstanceType> => {
+  const tokenRegistry = await TradeTrustErc721Factory.connect(address, wallet);
+  const titleEscrowAddress = await tokenRegistry.ownerOf(tokenId);
+  const titleEscrow = await TitleEscrowFactory.connect(titleEscrowAddress, wallet);
+  return titleEscrow;
+};
+
+interface validateEndorseOwnerArgs {
+  newHolder: string;
+  newOwner: string;
+  titleEscrow: TitleEscrowInstanceType;
+}
+export const validateEndorseOwner = async ({
```

Comment:
why not go a bit further and validate the state transition pre-conditions too?

Reply:
just concerned that it will bloat the code here since Infura will throw an error if the pre-conditions are wrong (just that its a vague error only CANNOT ESTIMATE GAS blah blah)

Reply:
i think it might be possible to extract the error string, check https://gist.github.com/gluk64/fdea559472d957f1138ed93bcbc6f78a#file-reason-js

Reply:
i think i'll open another PR for this, doesn't seem to be a quick fix here, I cant seem to get the `transactionHash` required to even kickstart the decoding.

```
 <ref *1> {
  data: '0x5a87db7500000000000000000000000011959c7272150a50603610052d278422bb5ae57f',
  to: Promise { '0xdE3F3CEe55FF7163DeF442aC0eF932884D01A33d' },
  from: '0x5B12Dc6A5209AE609c4BC72045BA09Cf5153d8CA',
  type: 2,
  maxFeePerGas: BigNumber { _hex: '0x773593f6', _isBigNumber: true },
  maxPriorityFeePerGas: BigNumber { _hex: '0x773593f6', _isBigNumber: true },
  nonce: Promise { 2 },
  gasLimit: Promise {
    <rejected> Error: cannot estimate gas; transaction may fail or may require manual gas limit (error={"reason":"cannot estimate gas; transaction may fail or may require manual gas limit","code":"UNPREDICTABLE_GAS_LIMIT","method":"estimateGas","transaction":{"from":"0x5B12Dc6A5209AE609c4BC72045BA09Cf5153d8CA","to":"0xdE3F3CEe55FF7163DeF442aC0eF932884D01A33d","data":"0x5a87db7500000000000000000000000011959c7272150a50603610052d278422bb5ae57f","type":2,"accessList":null}}, tx={"data":"0x5a87db7500000000000000000000000011959c7272150a50603610052d278422bb5ae57f","to":{},"from":"0x5B12Dc6A5209AE609c4BC72045BA09Cf5153d8CA","type":2,"maxFeePerGas":{"type":"BigNumber","hex":"0x773593f6"},"maxPriorityFeePerGas":{"type":"BigNumber","hex":"0x773593f6"},"nonce":{},"gasLimit":{},"chainId":{}}, code=UNPREDICTABLE_GAS_LIMIT, version=abstract-signer/5.4.0)
        at Logger.makeError (/Users/zhaoloong/heap/open-xxx/open-attestation-cli/node_modules/@ethersproject/wallet/node_modules/@ethersproject/logger/lib/index.js:187:21)
        at Logger.throwError (/Users/zhaoloong/heap/open-xxx/open-attestation-cli/node_modules/@ethersproject/wallet/node_modules/@ethersproject/logger/lib/index.js:196:20)
        at /Users/zhaoloong/heap/open-xxx/open-attestation-cli/node_modules/@ethersproject/wallet/node_modules/@ethersproject/abstract-signer/lib/index.js:353:47
        at processTicksAndRejections (internal/process/task_queues.js:93:5)
        at async Promise.all (index 7) {
      reason: 'cannot estimate gas; transaction may fail or may require manual gas limit',
      code: 'UNPREDICTABLE_GAS_LIMIT',
      error: [Error],
      tx: [Circular *1]
    }
  },
  chainId: Promise { 3 }
}
```

this is from the caught error.

thinking of being lazy and getting a library to help with the decoding https://github.com/authereum/eth-revert-reason

Reply:
feel free, not a blocker for me

ps: i fully endorse laziness 🤣